enhancements/cli-in-cluster-management.md

```diff
@@ -0,0 +1,164 @@
+---
+title: cli-in-cluster-management
+authors:
+  - "@sallyom"
+reviewers:
+  - "@soltysh"
+approvers:
+  - "@soltysh"
+creation-date: 2019-12-03
+last-updated: 2020-05-10
+status: implementable
+---
+
+# CLI In-Cluster Management
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This proposal is describing the mechanism for how authors of a Command Line Interface (CLI) such as oc, kubectl, odo, istio, tekton, or knative,
+can deliver tools to OpenShift clusters.  An in-cluster feature is needed to manage various CLIs available for OpenShift and related services.  The goal is for
+ a user to discover, install and upgrade tools easily and from a single location.
+
+Each component is responsible for building and publishing its artifacts and registering information regarding supplied binaries.
+Currently, that location is [index of /pub/openshift-v4/clients](https://mirror.openshift.com/pub/openshift-v4/clients/),
+and this makes it difficult for disconnected installations to mirror them.
+
+The CLI in-cluster manager will retrieve binaries from images that package a CLI's artifacts.  Through the use of local mirrored
+registries, disconnected environments will have access to OpenShift CLIs.  Currently, we provide disconnected environments 
```

Comment:
I don't know that it is required that binary extraction be done offline.  That is a "nice" feature we have but is not required.  I think this proposal could just as equally be "OCM and ACM provide a consistent mechanism for getting binary plugins, use ACM is if available".  The proposal needs to argue why a per-cluster approach is appropriate.

Reply:
@soltysh if installing disconnected is not a requirement, krew-index does not need to be served within cluster, right?  We would maintain a gh-repo like `openshift/oc-plugin-index` and configure oc with that index?

Reply:
As discussed yesterday on a planning, yeah we can go w/o disconnected support for starters, if the mechanism stays then adding it will be a simple switch.